// node-pg-migrate/dist/legacy/migration.js
import { getNumericPrefix, getSuffixFromFileName } from "./utils/fileNameUtils.js";
import { compareMigrationFileNames } from "./utils/comparators.js";
import { getMigrationTableSchema } from "./utils/getMigrationTableSchema.js";
import "./utils/index.js";
import { MigrationBuilder } from "./migrationBuilder.js";
import { glob } from "glob";
import { createReadStream, createWriteStream } from "node:fs";
import { mkdir, readdir } from "node:fs/promises";
import { basename, extname, join, resolve } from "node:path";
import { cwd } from "node:process";
//#region src/migration.ts
const FilenameFormat = Object.freeze({
	timestamp: "timestamp",
	utc: "utc",
	index: "index"
});
const SEPARATOR = "_";
/**
* Reads files from `dir`, sorts them and returns an array of their absolute paths.
* When not using globs, files are sorted by their numeric prefix values first. 17 digit numbers are interpreted as utc date and converted to the number representation of that date.
* Glob matches are sorted via String.localeCompare with ignored punctuation.
*
* @param dir The directory containing your migration files. This path is resolved from `cwd()`.
* Alternatively, provide a [glob](https://www.npmjs.com/package/glob) pattern or
* an array of glob patterns and set `options.useGlob = true`
*
* Note: enabling glob will read both, `dir` _and_ `options.ignorePattern` as glob patterns
* @param options
* @returns Array of absolute paths
*/
async function getMigrationFilePaths(dir, options = {}) {
	const { ignorePattern, useGlob = false, logger } = options;
	if (useGlob) return (await glob(dir, {
		ignore: ignorePattern,
		nodir: true,
		withFileTypes: true
	})).toSorted((a, b) => compareMigrationFileNames(a.name, b.name, logger)).map((pathScurry) => pathScurry.fullpath());
	if (Array.isArray(dir) || Array.isArray(ignorePattern)) throw new TypeError("Options \"dir\" and \"ignorePattern\" can only be arrays when \"useGlob\" is true");
	const ignoreRegexp = new RegExp(ignorePattern?.length ? `^${ignorePattern}$` : "^\\..*");
	return (await readdir(`${dir}/`, { withFileTypes: true })).filter((dirent) => (dirent.isFile() || dirent.isSymbolicLink()) && !ignoreRegexp.test(dirent.name)).toSorted((a, b) => compareMigrationFileNames(a.name, b.name, logger)).map((dirent) => resolve(dir, dirent.name));
}
async function getLastSuffix(dir, ignorePattern) {
	try {
		const files = await getMigrationFilePaths(dir, { ignorePattern });
		return files.length > 0 ? getSuffixFromFileName(files[files.length - 1]) : void 0;
	} catch {
		return;
	}
}
async function resolveSuffix(directory, options) {
	const { language, ignorePattern } = options;
	return language || await getLastSuffix(directory, ignorePattern) || "js";
}
var Migration = class Migration {
	/**
	* Get file prefix for a new migrations file
	*
	* @method Migration.getFilePrefix
	* @param filenameFormat Filename format
	* @param directory Migrations directory
	* @param [ignorePattern] Glob ignore pattern
	* @returns string New file prefix
	*/
	static async getFilePrefix(filenameFormat, directory, ignorePattern) {
		if (filenameFormat === FilenameFormat.index) {
			const filePaths = await getMigrationFilePaths(directory, {
				ignorePattern,
				useGlob: /\*/.test(directory) || /\*/.test(ignorePattern || "")
			});
			let lastPrefix = 0;
			let prefixLength = 0;
			for (const filenamePath of filePaths) {
				const filename = basename(filenamePath);
				if (!/^\d+\D/.test(filename)) throw new Error(`Cannot deduce index for previously created file "${filenamePath}"`);
				const prefix = filename.split(/\D/)[0] ?? "";
				prefixLength = Math.max(prefixLength, prefix.length);
				lastPrefix = Math.max(lastPrefix, getNumericPrefix(filename));
			}
			return `${lastPrefix + 1}`.padStart(prefixLength || 4, "0");
		}
		return filenameFormat === FilenameFormat.utc ? (/* @__PURE__ */ new Date()).toISOString().replace(/\D/g, "") : Date.now().toString();
	}
	static async create(name, directory, options = {}) {
		const { filenameFormat = FilenameFormat.timestamp, ignorePattern } = options;
		await mkdir(directory, { recursive: true });
		const prefix = await Migration.getFilePrefix(filenameFormat, directory, ignorePattern);
		const templateFileName = "templateFileName" in options ? resolve(cwd(), options.templateFileName) : join(import.meta.dirname, "..", "..", "templates", `migration-template.${await resolveSuffix(directory, options)}`);
		const newFile = join(directory, `${prefix}${SEPARATOR}${name}.${getSuffixFromFileName(templateFileName)}`);
		await new Promise((resolve, reject) => {
			createReadStream(templateFileName).pipe(createWriteStream(newFile)).on("close", resolve).on("error", reject);
		});
		return newFile;
	}
	db;
	path;
	name;
	timestamp;
	up;
	down;
	options;
	typeShorthands;
	logger;
	constructor(db, migrationPath, { up, down }, options, typeShorthands, logger = console) {
		this.db = db;
		this.path = migrationPath;
		this.name = basename(migrationPath, extname(migrationPath));
		this.up = up;
		this.down = down;
		this.options = options;
		this.typeShorthands = typeShorthands;
		this.logger = logger;
		try {
			this.timestamp = getNumericPrefix(this.name, logger);
		} catch {
			this.timestamp = 0;
		}
	}
	_getMarkAsRun(action) {
		const schema = getMigrationTableSchema(this.options);
		const { migrationsTable } = this.options;
		const { name } = this;
		switch (action) {
			case this.down:
				this.logger.info(`### MIGRATION ${this.name} (DOWN) ###`);
				return `DELETE FROM "${schema}"."${migrationsTable}" WHERE name='${name}';`;
			case this.up:
				this.logger.info(`### MIGRATION ${this.name} (UP) ###`);
				return `INSERT INTO "${schema}"."${migrationsTable}" (name, run_on) VALUES ('${name}', NOW());`;
			default: throw new Error("Unknown direction");
		}
	}
	async _apply(action, pgm) {
		if (action.length === 2) await new Promise((resolve) => {
			action(pgm, resolve);
		});
		else await action(pgm);
		const sqlSteps = pgm.getSqlSteps();
		sqlSteps.push(this._getMarkAsRun(action));
		if (!this.options.singleTransaction && pgm.isUsingTransaction()) {
			sqlSteps.unshift("BEGIN;");
			sqlSteps.push("COMMIT;");
		} else if (this.options.singleTransaction && !pgm.isUsingTransaction()) {
			this.logger.warn("#> WARNING: Need to break single transaction! <");
			sqlSteps.unshift("COMMIT;");
			sqlSteps.push("BEGIN;");
		} else if (!this.options.singleTransaction || !pgm.isUsingTransaction()) this.logger.warn("#> WARNING: This migration is not wrapped in a transaction! <");
		if (typeof this.logger.debug === "function") this.logger.debug(`${sqlSteps.join("\n")}\n\n`);
		return sqlSteps.reduce((promise, sql) => promise.then(() => this.options.dryRun || this.db.query(sql)), Promise.resolve());
	}
	_getAction(direction) {
		if (direction === "down" && this.down === void 0) this.down = this.up;
		const action = this[direction];
		if (action === false) throw new Error(`User has disabled ${direction} migration on file: ${this.name}`);
		if (typeof action !== "function") throw new Error(`Unknown value for direction: ${direction}. Is the migration ${this.name} exporting a '${direction}' function?`);
		return action;
	}
	apply(direction) {
		const pgm = new MigrationBuilder(this.db, this.typeShorthands, Boolean(this.options.decamelize), this.logger);
		const action = this._getAction(direction);
		if (this.down === this.up) pgm.enableReverseMode();
		return this._apply(action, pgm);
	}
	markAsRun(direction) {
		return this.db.query(this._getMarkAsRun(this._getAction(direction)));
	}
};
//#endregion
export { FilenameFormat, Migration, getMigrationFilePaths };
