// node-pg-migrate/dist/legacy/migrationBuilder.js
import { dropCast } from "./operations/casts/dropCast.js";
import { createCast } from "./operations/casts/createCast.js";
import { PgLiteral } from "./utils/PgLiteral.js";
import { createSchemalize } from "./utils/createSchemalize.js";
import "./utils/index.js";
import { alterDomain } from "./operations/domains/alterDomain.js";
import { dropDomain } from "./operations/domains/dropDomain.js";
import { createDomain } from "./operations/domains/createDomain.js";
import { renameDomain } from "./operations/domains/renameDomain.js";
import "./operations/domains/index.js";
import { dropExtension } from "./operations/extensions/dropExtension.js";
import { createExtension } from "./operations/extensions/createExtension.js";
import "./operations/extensions/index.js";
import { dropFunction } from "./operations/functions/dropFunction.js";
import { createFunction } from "./operations/functions/createFunction.js";
import { renameFunction } from "./operations/functions/renameFunction.js";
import "./operations/functions/index.js";
import { revokeOnSchemas } from "./operations/grants/revokeOnSchemas.js";
import { grantOnSchemas } from "./operations/grants/grantOnSchemas.js";
import { revokeOnTables } from "./operations/grants/revokeOnTables.js";
import { grantOnTables } from "./operations/grants/grantOnTables.js";
import { revokeRoles } from "./operations/grants/revokeRoles.js";
import { grantRoles } from "./operations/grants/grantRoles.js";
import "./operations/grants/index.js";
import { dropIndex } from "./operations/indexes/dropIndex.js";
import { createIndex } from "./operations/indexes/createIndex.js";
import { renameIndex } from "./operations/indexes/renameIndex.js";
import "./operations/indexes/index.js";
import { alterMaterializedView } from "./operations/materializedViews/alterMaterializedView.js";
import { dropMaterializedView } from "./operations/materializedViews/dropMaterializedView.js";
import { createMaterializedView } from "./operations/materializedViews/createMaterializedView.js";
import { refreshMaterializedView } from "./operations/materializedViews/refreshMaterializedView.js";
import { renameMaterializedView } from "./operations/materializedViews/renameMaterializedView.js";
import { renameMaterializedViewColumn } from "./operations/materializedViews/renameMaterializedViewColumn.js";
import "./operations/materializedViews/index.js";
import { removeFromOperatorFamily } from "./operations/operators/removeFromOperatorFamily.js";
import { addToOperatorFamily } from "./operations/operators/addToOperatorFamily.js";
import { dropOperator } from "./operations/operators/dropOperator.js";
import { createOperator } from "./operations/operators/createOperator.js";
import { dropOperatorClass } from "./operations/operators/dropOperatorClass.js";
import { createOperatorClass } from "./operations/operators/createOperatorClass.js";
import { dropOperatorFamily } from "./operations/operators/dropOperatorFamily.js";
import { createOperatorFamily } from "./operations/operators/createOperatorFamily.js";
import { renameOperatorClass } from "./operations/operators/renameOperatorClass.js";
import { renameOperatorFamily } from "./operations/operators/renameOperatorFamily.js";
import "./operations/operators/index.js";
import { alterPolicy } from "./operations/policies/alterPolicy.js";
import { dropPolicy } from "./operations/policies/dropPolicy.js";
import { createPolicy } from "./operations/policies/createPolicy.js";
import { renamePolicy } from "./operations/policies/renamePolicy.js";
import "./operations/policies/index.js";
import { alterRole } from "./operations/roles/alterRole.js";
import { dropRole } from "./operations/roles/dropRole.js";
import { createRole } from "./operations/roles/createRole.js";
import { renameRole } from "./operations/roles/renameRole.js";
import "./operations/roles/index.js";
import { dropSchema } from "./operations/schemas/dropSchema.js";
import { createSchema } from "./operations/schemas/createSchema.js";
import { renameSchema } from "./operations/schemas/renameSchema.js";
import { alterSequence } from "./operations/sequences/alterSequence.js";
import { dropSequence } from "./operations/sequences/dropSequence.js";
import { createSequence } from "./operations/sequences/createSequence.js";
import { renameSequence } from "./operations/sequences/renameSequence.js";
import "./operations/sequences/index.js";
import { sql } from "./operations/sql.js";
import { dropColumns } from "./operations/tables/dropColumns.js";
import { addColumns } from "./operations/tables/addColumns.js";
import { dropConstraint } from "./operations/tables/dropConstraint.js";
import { addConstraint } from "./operations/tables/addConstraint.js";
import { alterColumn } from "./operations/tables/alterColumn.js";
import { alterTable } from "./operations/tables/alterTable.js";
import { dropTable } from "./operations/tables/dropTable.js";
import { createTable } from "./operations/tables/createTable.js";
import { renameColumn } from "./operations/tables/renameColumn.js";
import { renameConstraint } from "./operations/tables/renameConstraint.js";
import { renameTable } from "./operations/tables/renameTable.js";
import "./operations/tables/index.js";
import { dropTrigger } from "./operations/triggers/dropTrigger.js";
import { createTrigger } from "./operations/triggers/createTrigger.js";
import { renameTrigger } from "./operations/triggers/renameTrigger.js";
import "./operations/triggers/index.js";
import { dropTypeAttribute } from "./operations/types/dropTypeAttribute.js";
import { addTypeAttribute } from "./operations/types/addTypeAttribute.js";
import { addTypeValue } from "./operations/types/addTypeValue.js";
import { dropType } from "./operations/types/dropType.js";
import { createType } from "./operations/types/createType.js";
import { renameType } from "./operations/types/renameType.js";
import { renameTypeAttribute } from "./operations/types/renameTypeAttribute.js";
import { renameTypeValue } from "./operations/types/renameTypeValue.js";
import { setTypeAttribute } from "./operations/types/setTypeAttribute.js";
import "./operations/types/index.js";
import { alterView } from "./operations/views/alterView.js";
import { alterViewColumn } from "./operations/views/alterViewColumn.js";
import { dropView } from "./operations/views/dropView.js";
import { createView } from "./operations/views/createView.js";
import { renameView } from "./operations/views/renameView.js";
import "./operations/views/index.js";
//#region src/migrationBuilder.ts
/**
* Export "PgLiteral" of module "src/utils/PgLiteral.ts" was reexported through module "src/utils/index.ts"
* while both modules are dependencies of each other and will end up in different chunks by current Rollup settings.
* This scenario is not well supported at the moment as it will produce a circular dependency between chunks
* and will likely lead to broken execution order.
* Either change the import in "src/migrationBuilder.ts" to point directly to the exporting module
* or reconfigure "output.manualChunks" to ensure these modules end up in the same chunk.
*/
var MigrationBuilder = class {
	/**
	* Install an extension.
	*
	* @alias addExtension
	*
	* @see https://www.postgresql.org/docs/current/sql-createextension.html
	*/
	createExtension;
	/**
	* Remove an extension.
	*
	* @see https://www.postgresql.org/docs/current/sql-dropextension.html
	*/
	dropExtension;
	/**
	* Install an extension.
	*
	* @alias createExtension
	*
	* @see https://www.postgresql.org/docs/current/sql-createextension.html
	*/
	addExtension;
	/**
	* Define a new table.
	*
	* @see https://www.postgresql.org/docs/current/sql-createtable.html
	*/
	createTable;
	/**
	* Remove a table.
	*
	* @see https://www.postgresql.org/docs/current/sql-droptable.html
	*/
	dropTable;
	/**
	* Rename a table.
	*
	* @see https://www.postgresql.org/docs/current/sql-altertable.html
	*/
	renameTable;
	/**
	* Change the definition of a table.
	*
	* @see https://www.postgresql.org/docs/current/sql-altertable.html
	*/
	alterTable;
	/**
	* Add columns to a table.
	*
	* @alias addColumn
	*
	* @see https://www.postgresql.org/docs/current/sql-altertable.html
	*/
	addColumns;
	/**
	* Remove columns from a table.
	*
	* @alias dropColumn
	*
	* @see https://www.postgresql.org/docs/current/sql-altertable.html
	*/
	dropColumns;
	/**
	* Rename a column.
	*
	* @see https://www.postgresql.org/docs/current/sql-altertable.html
	*/
	renameColumn;
	/**
	* Change the definition of a column.
	*
	* @see https://www.postgresql.org/docs/current/sql-altertable.html
	*/
	alterColumn;
	/**
	* Add a column to a table.
	*
	* @alias addColumns
	*
	* @see https://www.postgresql.org/docs/current/sql-altertable.html
	*/
	addColumn;
	/**
	* Remove a column from a table.
	*
	* @alias dropColumns
	*
	* @see https://www.postgresql.org/docs/current/sql-altertable.html
	*/
	dropColumn;
	/**
	* Add a constraint to a table.
	*
	* @alias createConstraint
	*
	* @see https://www.postgresql.org/docs/current/sql-altertable.html
	*/
	addConstraint;
	/**
	* Remove a constraint from a table.
	*
	* @see https://www.postgresql.org/docs/current/sql-altertable.html
	*/
	dropConstraint;
	/**
	* Rename a constraint.
	*
	* @see https://www.postgresql.org/docs/current/sql-altertable.html
	*/
	renameConstraint;
	/**
	* Add a constraint to a table.
	*
	* @alias addConstraint
	*
	* @see https://www.postgresql.org/docs/current/sql-altertable.html
	*/
	createConstraint;
	/**
	* Define a new data type.
	*
	* @alias addType
	*
	* @see https://www.postgresql.org/docs/current/sql-createtype.html
	*/
	createType;
	/**
	* Remove a data type.
	*
	* @see https://www.postgresql.org/docs/current/sql-droptype.html
	*/
	dropType;
	/**
	* Define a new data type.
	*
	* @alias createType
	*
	* @see https://www.postgresql.org/docs/current/sql-createtype.html
	*/
	addType;
	/**
	* Rename a data type.
	*
	* @see https://www.postgresql.org/docs/current/sql-altertype.html
	*/
	renameType;
	/**
	* Rename a data type attribute.
	*
	* @see https://www.postgresql.org/docs/current/sql-altertype.html
	*/
	renameTypeAttribute;
	/**
	* Rename a data type value.
	*
	* @see https://www.postgresql.org/docs/current/sql-altertype.html
	*/
	renameTypeValue;
	/**
	* Add an attribute to a data type.
	*
	* @see https://www.postgresql.org/docs/current/sql-altertype.html
	*/
	addTypeAttribute;
	/**
	* Remove an attribute from a data type.
	*
	* @see https://www.postgresql.org/docs/current/sql-altertype.html
	*/
	dropTypeAttribute;
	/**
	* Set an attribute of a data type.
	*
	* @see https://www.postgresql.org/docs/current/sql-altertype.html
	*/
	setTypeAttribute;
	/**
	* Add a value to a data type.
	*
	* @see https://www.postgresql.org/docs/current/sql-altertype.html
	*/
	addTypeValue;
	/**
	* Define a new index.
	*
	* @alias addIndex
	*
	* @see https://www.postgresql.org/docs/current/sql-createindex.html
	*/
	createIndex;
	/**
	* Remove an index.
	*
	* @see https://www.postgresql.org/docs/current/sql-dropindex.html
	*/
	dropIndex;
	/**
	* Define a new index.
	*
	* @alias createIndex
	*
	* @see https://www.postgresql.org/docs/current/sql-createindex.html
	*/
	addIndex;
	/**
	* Rename an index.
	*
	* @see https://www.postgresql.org/docs/current/sql-alterindex.html
	*/
	renameIndex;
	/**
	* Define a new database role.
	*
	* @see https://www.postgresql.org/docs/current/sql-createrole.html
	*/
	createRole;
	/**
	* Remove a database role.
	*
	* @see https://www.postgresql.org/docs/current/sql-droprole.html
	*/
	dropRole;
	/**
	* Change a database role.
	*
	* @see https://www.postgresql.org/docs/current/sql-alterrole.html
	*/
	alterRole;
	/**
	* Rename a database role.
	*
	* @see https://www.postgresql.org/docs/current/sql-alterrole.html
	*/
	renameRole;
	/**
	* Define a new function.
	*
	* @see https://www.postgresql.org/docs/current/sql-createfunction.html
	*/
	createFunction;
	/**
	* Remove a function.
	*
	* @see https://www.postgresql.org/docs/current/sql-dropfunction.html
	*/
	dropFunction;
	/**
	* Rename a function.
	*
	* @see https://www.postgresql.org/docs/current/sql-alterfunction.html
	*/
	renameFunction;
	/**
	* Define a new trigger.
	*
	* @see https://www.postgresql.org/docs/current/sql-createtrigger.html
	*/
	createTrigger;
	/**
	* Remove a trigger.
	*
	* @see https://www.postgresql.org/docs/current/sql-droptrigger.html
	*/
	dropTrigger;
	/**
	* Rename a trigger.
	*
	* @see https://www.postgresql.org/docs/current/sql-altertrigger.html
	*/
	renameTrigger;
	/**
	* Define a new schema.
	*
	* @see https://www.postgresql.org/docs/current/sql-createschema.html
	*/
	createSchema;
	/**
	* Remove a schema.
	*
	* @see https://www.postgresql.org/docs/current/sql-dropschema.html
	*/
	dropSchema;
	/**
	* Rename a schema.
	*
	* @see https://www.postgresql.org/docs/current/sql-alterschema.html
	*/
	renameSchema;
	/**
	* Define a new domain.
	*
	* @see https://www.postgresql.org/docs/current/sql-createdomain.html
	*/
	createDomain;
	/**
	* Remove a domain.
	*
	* @see https://www.postgresql.org/docs/current/sql-dropdomain.html
	*/
	dropDomain;
	/**
	* Change the definition of a domain.
	*
	* @see https://www.postgresql.org/docs/current/sql-alterdomain.html
	*/
	alterDomain;
	/**
	* Rename a domain.
	*
	* @see https://www.postgresql.org/docs/current/sql-alterdomain.html
	*/
	renameDomain;
	/**
	* Define a new sequence generator.
	*
	* @see https://www.postgresql.org/docs/current/sql-createsequence.html
	*/
	createSequence;
	/**
	* Remove a sequence.
	*
	* @see https://www.postgresql.org/docs/current/sql-dropsequence.html
	*/
	dropSequence;
	/**
	* Change the definition of a sequence generator.
	*
	* @see https://www.postgresql.org/docs/current/sql-altersequence.html
	*/
	alterSequence;
	/**
	* Rename a sequence.
	*
	* @see https://www.postgresql.org/docs/current/sql-altersequence.html
	*/
	renameSequence;
	/**
	* Define a new operator.
	*
	* @see https://www.postgresql.org/docs/current/sql-createoperator.html
	*/
	createOperator;
	/**
	* Remove an operator.
	*
	* @see https://www.postgresql.org/docs/current/sql-dropoperator.html
	*/
	dropOperator;
	/**
	* Define a new operator class.
	*
	* @see https://www.postgresql.org/docs/current/sql-createopclass.html
	*/
	createOperatorClass;
	/**
	* Remove an operator class.
	*
	* @see https://www.postgresql.org/docs/current/sql-dropopclass.html
	*/
	dropOperatorClass;
	/**
	* Rename an operator class.
	*
	* @see https://www.postgresql.org/docs/current/sql-alteropclass.html
	*/
	renameOperatorClass;
	/**
	* Define a new operator family.
	*
	* @see https://www.postgresql.org/docs/current/sql-createopfamily.html
	*/
	createOperatorFamily;
	/**
	* Remove an operator family.
	*
	* @see https://www.postgresql.org/docs/current/sql-dropopfamily.html
	*/
	dropOperatorFamily;
	/**
	* Rename an operator family.
	*
	* @see https://www.postgresql.org/docs/current/sql-alteropfamily.html
	*/
	renameOperatorFamily;
	/**
	* Add an operator to an operator family.
	*
	* @see https://www.postgresql.org/docs/current/sql-alteropfamily.html
	*/
	addToOperatorFamily;
	/**
	* Remove an operator from an operator family.
	*
	* @see https://www.postgresql.org/docs/current/sql-alteropfamily.html
	*/
	removeFromOperatorFamily;
	/**
	* Define a new row-level security policy for a table.
	*
	* @see https://www.postgresql.org/docs/current/sql-createpolicy.html
	*/
	createPolicy;
	/**
	* Remove a row-level security policy from a table.
	*
	* @see https://www.postgresql.org/docs/current/sql-droppolicy.html
	*/
	dropPolicy;
	/**
	* Change the definition of a row-level security policy.
	*
	* @see https://www.postgresql.org/docs/current/sql-alterpolicy.html
	*/
	alterPolicy;
	/**
	* Rename a row-level security policy.
	*
	* @see https://www.postgresql.org/docs/current/sql-alterpolicy.html
	*/
	renamePolicy;
	/**
	* Define a new view.
	*
	* @see https://www.postgresql.org/docs/current/sql-createview.html
	*/
	createView;
	/**
	* Remove a view.
	*
	* @see https://www.postgresql.org/docs/current/sql-dropview.html
	*/
	dropView;
	/**
	* Change the definition of a view.
	*
	* @see https://www.postgresql.org/docs/current/sql-alterview.html
	*/
	alterView;
	/**
	* Change the definition of a view column.
	*
	* @see https://www.postgresql.org/docs/current/sql-alterview.html
	*/
	alterViewColumn;
	/**
	* Rename a view.
	*
	* @see https://www.postgresql.org/docs/current/sql-alterview.html
	*/
	renameView;
	/**
	* Define a new materialized view.
	*
	* @see https://www.postgresql.org/docs/current/sql-creatematerializedview.html
	*/
	createMaterializedView;
	/**
	* Remove a materialized view.
	*
	* @see https://www.postgresql.org/docs/current/sql-dropmaterializedview.html
	*/
	dropMaterializedView;
	/**
	* Change the definition of a materialized view.
	*
	* @see https://www.postgresql.org/docs/current/sql-altermaterializedview.html
	*/
	alterMaterializedView;
	/**
	* Rename a materialized view.
	*
	* @see https://www.postgresql.org/docs/current/sql-altermaterializedview.html
	*/
	renameMaterializedView;
	/**
	* Rename a materialized view column.
	*
	* @see https://www.postgresql.org/docs/current/sql-altermaterializedview.html
	*/
	renameMaterializedViewColumn;
	/**
	* Replace the contents of a materialized view.
	*
	* @see https://www.postgresql.org/docs/current/sql-refreshmaterializedview.html
	*/
	refreshMaterializedView;
	/**
	* Define access privileges.
	*
	* @see https://www.postgresql.org/docs/current/sql-grant.html
	*/
	grantRoles;
	/**
	* Remove access privileges.
	*
	* @see https://www.postgresql.org/docs/current/sql-revoke.html
	*/
	revokeRoles;
	/**
	* Define access privileges.
	*
	* @see https://www.postgresql.org/docs/current/sql-grant.html
	*/
	grantOnSchemas;
	/**
	* Remove access privileges.
	*
	* @see https://www.postgresql.org/docs/current/sql-revoke.html
	*/
	revokeOnSchemas;
	/**
	* Define access privileges.
	*
	* @see https://www.postgresql.org/docs/current/sql-grant.html
	*/
	grantOnTables;
	/**
	* Remove access privileges.
	*
	* @see https://www.postgresql.org/docs/current/sql-revoke.html
	*/
	revokeOnTables;
	/**
	* Define a new cast.
	*
	* @see https://www.postgresql.org/docs/current/sql-createcast.html
	*/
	createCast;
	/**
	* Remove a cast.
	*
	* @see https://www.postgresql.org/docs/current/sql-dropcast.html
	*/
	dropCast;
	/**
	* Run raw SQL, with some optional _[very basic](http://mir.aculo.us/2011/03/09/little-helpers-a-tweet-sized-javascript-templating-engine/)_ mustache templating.
	*
	* This is a low-level operation, and you should use the higher-level operations whenever possible.
	*
	* @param sql SQL query to run.
	* @param args Optional `key/val` of arguments to replace.
	*
	* @see https://www.postgresql.org/docs/current/sql-commands.html
	*/
	sql;
	/**
	* Inserts raw string, **which is not escaped**.
	*
	* @param sql String to **not escaped**.
	*
	* @example
	* { default: pgm.func('CURRENT_TIMESTAMP') }
	*/
	func;
	/**
	* The `db` client instance.
	*
	* Can be used to run queries directly.
	*/
	db;
	_steps;
	_REVERSE_MODE;
	_useTransaction;
	constructor(db, typeShorthands, shouldDecamelize, logger) {
		this._steps = [];
		this._REVERSE_MODE = false;
		this._useTransaction = true;
		const wrap = (operation) => (...args) => {
			if (this._REVERSE_MODE) {
				if (typeof operation.reverse !== "function") {
					const name = `pgm.${operation.name}()`;
					throw new Error(`Impossible to automatically infer down migration for "${name}"`);
				}
				this._steps = this._steps.concat(operation.reverse(...args));
			} else this._steps = this._steps.concat(operation(...args));
		};
		const options = {
			typeShorthands,
			schemalize: createSchemalize({
				shouldDecamelize,
				shouldQuote: false
			}),
			literal: createSchemalize({
				shouldDecamelize,
				shouldQuote: true
			}),
			logger
		};
		this.createExtension = wrap(createExtension(options));
		this.dropExtension = wrap(dropExtension(options));
		this.addExtension = this.createExtension;
		this.createTable = wrap(createTable(options));
		this.dropTable = wrap(dropTable(options));
		this.renameTable = wrap(renameTable(options));
		this.alterTable = wrap(alterTable(options));
		this.addColumns = wrap(addColumns(options));
		this.dropColumns = wrap(dropColumns(options));
		this.renameColumn = wrap(renameColumn(options));
		this.alterColumn = wrap(alterColumn(options));
		this.addColumn = this.addColumns;
		this.dropColumn = this.dropColumns;
		this.addConstraint = wrap(addConstraint(options));
		this.dropConstraint = wrap(dropConstraint(options));
		this.renameConstraint = wrap(renameConstraint(options));
		this.createConstraint = this.addConstraint;
		this.createType = wrap(createType(options));
		this.dropType = wrap(dropType(options));
		this.addType = this.createType;
		this.renameType = wrap(renameType(options));
		this.renameTypeAttribute = wrap(renameTypeAttribute(options));
		this.renameTypeValue = wrap(renameTypeValue(options));
		this.addTypeAttribute = wrap(addTypeAttribute(options));
		this.dropTypeAttribute = wrap(dropTypeAttribute(options));
		this.setTypeAttribute = wrap(setTypeAttribute(options));
		this.addTypeValue = wrap(addTypeValue(options));
		this.createIndex = wrap(createIndex(options));
		this.dropIndex = wrap(dropIndex(options));
		this.addIndex = this.createIndex;
		this.renameIndex = wrap(renameIndex(options));
		this.createRole = wrap(createRole(options));
		this.dropRole = wrap(dropRole(options));
		this.alterRole = wrap(alterRole(options));
		this.renameRole = wrap(renameRole(options));
		this.createFunction = wrap(createFunction(options));
		this.dropFunction = wrap(dropFunction(options));
		this.renameFunction = wrap(renameFunction(options));
		this.createTrigger = wrap(createTrigger(options));
		this.dropTrigger = wrap(dropTrigger(options));
		this.renameTrigger = wrap(renameTrigger(options));
		this.createSchema = wrap(createSchema(options));
		this.dropSchema = wrap(dropSchema(options));
		this.renameSchema = wrap(renameSchema(options));
		this.createDomain = wrap(createDomain(options));
		this.dropDomain = wrap(dropDomain(options));
		this.alterDomain = wrap(alterDomain(options));
		this.renameDomain = wrap(renameDomain(options));
		this.createSequence = wrap(createSequence(options));
		this.dropSequence = wrap(dropSequence(options));
		this.alterSequence = wrap(alterSequence(options));
		this.renameSequence = wrap(renameSequence(options));
		this.createOperator = wrap(createOperator(options));
		this.dropOperator = wrap(dropOperator(options));
		this.createOperatorClass = wrap(createOperatorClass(options));
		this.dropOperatorClass = wrap(dropOperatorClass(options));
		this.renameOperatorClass = wrap(renameOperatorClass(options));
		this.createOperatorFamily = wrap(createOperatorFamily(options));
		this.dropOperatorFamily = wrap(dropOperatorFamily(options));
		this.renameOperatorFamily = wrap(renameOperatorFamily(options));
		this.addToOperatorFamily = wrap(addToOperatorFamily(options));
		this.removeFromOperatorFamily = wrap(removeFromOperatorFamily(options));
		this.createPolicy = wrap(createPolicy(options));
		this.dropPolicy = wrap(dropPolicy(options));
		this.alterPolicy = wrap(alterPolicy(options));
		this.renamePolicy = wrap(renamePolicy(options));
		this.createView = wrap(createView(options));
		this.dropView = wrap(dropView(options));
		this.alterView = wrap(alterView(options));
		this.alterViewColumn = wrap(alterViewColumn(options));
		this.renameView = wrap(renameView(options));
		this.createMaterializedView = wrap(createMaterializedView(options));
		this.dropMaterializedView = wrap(dropMaterializedView(options));
		this.alterMaterializedView = wrap(alterMaterializedView(options));
		this.renameMaterializedView = wrap(renameMaterializedView(options));
		this.renameMaterializedViewColumn = wrap(renameMaterializedViewColumn(options));
		this.refreshMaterializedView = wrap(refreshMaterializedView(options));
		this.grantRoles = wrap(grantRoles(options));
		this.revokeRoles = wrap(revokeRoles(options));
		this.grantOnSchemas = wrap(grantOnSchemas(options));
		this.revokeOnSchemas = wrap(revokeOnSchemas(options));
		this.grantOnTables = wrap(grantOnTables(options));
		this.revokeOnTables = wrap(revokeOnTables(options));
		this.createCast = wrap(createCast(options));
		this.dropCast = wrap(dropCast(options));
		this.sql = wrap(sql(options));
		this.func = PgLiteral.create;
		const wrapDB = (operation) => (...args) => {
			if (this._REVERSE_MODE) throw new Error("Impossible to automatically infer down migration");
			return operation(...args);
		};
		this.db = {
			query: wrapDB(db.query),
			select: wrapDB(db.select)
		};
	}
	/**
	* Run the reverse of the migration. Useful for creating a new migration that
	* reverts a previous migration.
	*/
	enableReverseMode() {
		this._REVERSE_MODE = true;
		return this;
	}
	/**
	* By default, all migrations are run in one transaction, but some DB
	* operations like add type value (`pgm.addTypeValue`) does not work if the
	* type is not created in the same transaction.
	* e.g. if it is created in previous migration. You need to run specific
	* migration outside a transaction (`pgm.noTransaction`).
	* Be aware that this means that you can have some migrations applied and some
	* not applied, if there is some error during migrating (leading to `ROLLBACK`).
	*/
	noTransaction() {
		this._useTransaction = false;
		return this;
	}
	isUsingTransaction() {
		return this._useTransaction;
	}
	getSql() {
		return `${this.getSqlSteps().join("\n")}\n`;
	}
	getSqlSteps() {
		return this._REVERSE_MODE ? this._steps.toReversed() : this._steps;
	}
};
//#endregion
export { MigrationBuilder };
