// node-pg-migrate/dist/legacy/runner.js
import { db } from "./db.js";
import { createSchemalize } from "./utils/createSchemalize.js";
import { getMigrationTableSchema } from "./utils/getMigrationTableSchema.js";
import { getSchemas } from "./utils/getSchemas.js";
import "./utils/index.js";
import { Migration, getMigrationFilePaths } from "./migration.js";
import { loadMigrationUnits } from "./migrationLoader.js";
//#region src/runner.ts
/**
* Random but well-known identifier shared by all instances of `node-pg-migrate`.
*/
const PG_MIGRATE_LOCK_ID = 7241865325823964;
const idColumn = "id";
const nameColumn = "name";
const runOnColumn = "run_on";
async function loadMigrations(db, options, logger) {
	try {
		let shorthands = {};
		const absoluteFilePaths = await getMigrationFilePaths(options.dir, {
			ignorePattern: options.ignorePattern,
			useGlob: options.useGlob,
			logger
		});
		const migrations = [];
		const migrationUnits = await loadMigrationUnits(options, absoluteFilePaths);
		for (const { id: filePath, actions } of migrationUnits) {
			shorthands = {
				...shorthands,
				...actions.shorthands
			};
			migrations.push(new Migration(db, filePath, actions, options, { ...shorthands }, logger));
		}
		return migrations;
	} catch (error) {
		if (error instanceof Error) throw new Error(`Error loading migration files: ${error.message}`, { cause: error });
		throw new Error("Error loading migration files: Unknown error", { cause: error });
	}
}
async function lock(db, lockValue = PG_MIGRATE_LOCK_ID, advisoryLockMode = "fail") {
	if (advisoryLockMode === "wait") {
		await db.query(`SELECT pg_advisory_lock(${lockValue})`);
		return;
	}
	const [result] = await db.select(`SELECT pg_try_advisory_lock(${lockValue}) AS "lockObtained"`);
	if (!result.lockObtained) throw new Error("Another migration is already running. Advisory lock mode is set to 'fail'.");
}
async function unlock(db, lockValue = PG_MIGRATE_LOCK_ID) {
	const [result] = await db.select(`SELECT pg_advisory_unlock(${lockValue}) AS "lockReleased"`);
	if (!result.lockReleased) throw new Error("Failed to release migration lock");
}
async function ensureMigrationsTable(db, options) {
	try {
		const schema = getMigrationTableSchema(options);
		const { migrationsTable } = options;
		const fullTableName = createSchemalize({
			shouldDecamelize: Boolean(options.decamelize),
			shouldQuote: true
		})({
			schema,
			name: migrationsTable
		});
		const migrationTables = await db.select(`SELECT table_name FROM information_schema.tables WHERE table_schema = '${schema}' AND table_name = '${migrationsTable}'`);
		if (migrationTables && migrationTables.length === 1) {
			const primaryKeyConstraints = await db.select(`SELECT constraint_name FROM information_schema.table_constraints WHERE table_schema = '${schema}' AND table_name = '${migrationsTable}' AND constraint_type = 'PRIMARY KEY'`);
			if (!primaryKeyConstraints || primaryKeyConstraints.length !== 1) await db.query(`ALTER TABLE ${fullTableName} ADD PRIMARY KEY (${idColumn})`);
		} else await db.query(`CREATE TABLE ${fullTableName} (${idColumn} SERIAL PRIMARY KEY, ${nameColumn} varchar(255) NOT NULL, ${runOnColumn} timestamp NOT NULL)`);
	} catch (error) {
		throw new Error(`Unable to ensure migrations table: ${error.stack}`);
	}
}
async function getRunMigrations(db, options) {
	const schema = getMigrationTableSchema(options);
	const { migrationsTable } = options;
	const fullTableName = createSchemalize({
		shouldDecamelize: Boolean(options.decamelize),
		shouldQuote: true
	})({
		schema,
		name: migrationsTable
	});
	return db.column(nameColumn, `SELECT ${nameColumn} FROM ${fullTableName} ORDER BY ${runOnColumn}, ${idColumn}`);
}
function getMigrationsToRun(options, runNames, migrations) {
	if (options.direction === "down") {
		const downMigrations = runNames.filter((migrationName) => !options.file || options.file === migrationName).map((migrationName) => migrations.find(({ name }) => name === migrationName) || migrationName);
		const { count = 1 } = options;
		const toRun = (options.timestamp ? downMigrations.filter((migration) => typeof migration === "object" && migration.timestamp >= count) : downMigrations.slice(-Math.abs(count))).toReversed();
		const deletedMigrations = toRun.filter((migration) => typeof migration === "string");
		if (deletedMigrations.length > 0) {
			const deletedMigrationsStr = deletedMigrations.join(", ");
			throw new Error(`Definitions of migrations ${deletedMigrationsStr} have been deleted.`);
		}
		return toRun;
	}
	const upMigrations = migrations.filter(({ name }) => !runNames.includes(name) && (!options.file || options.file === name));
	const { count = Number.POSITIVE_INFINITY } = options;
	return options.timestamp ? upMigrations.filter(({ timestamp }) => timestamp <= count) : upMigrations.slice(0, Math.abs(count));
}
function checkOrder(runNames, migrations) {
	const len = Math.min(runNames.length, migrations.length);
	for (let i = 0; i < len; i += 1) {
		const runName = runNames[i];
		const migrationName = migrations[i].name;
		if (runName !== migrationName) throw new Error(`Not run migration ${migrationName} is preceding already run migration ${runName}`);
	}
}
function runMigrations(toRun, method, direction) {
	return toRun.reduce((promise, migration) => promise.then(() => migration[method](direction)), Promise.resolve());
}
function getLogger(options) {
	const { log, logger, verbose } = options;
	let loggerObject = console;
	if (typeof logger === "object") loggerObject = logger;
	else if (typeof log === "function") loggerObject = {
		debug: log,
		info: log,
		warn: log,
		error: log
	};
	return verbose ? loggerObject : {
		debug: void 0,
		info: loggerObject.info.bind(loggerObject),
		warn: loggerObject.warn.bind(loggerObject),
		error: loggerObject.error.bind(loggerObject)
	};
}
async function runner(options) {
	const logger = getLogger(options);
	const connection = options.dbClient || options.databaseUrl;
	if (connection == null) throw new Error("You must provide either a databaseUrl or a dbClient");
	const db$1 = db(connection, logger);
	try {
		await db$1.createConnection();
		if (!options.noLock) await lock(db$1, options.lockValue, options.advisoryLockMode);
		if (options.schema) {
			const schemas = getSchemas(options.schema);
			if (options.createSchema) await Promise.all(schemas.map((schema) => db$1.query(`CREATE SCHEMA IF NOT EXISTS "${schema}"`)));
			await db$1.query(`SET search_path TO ${schemas.map((s) => `"${s}"`).join(", ")}`);
		}
		if (options.migrationsSchema && options.createMigrationsSchema) await db$1.query(`CREATE SCHEMA IF NOT EXISTS "${options.migrationsSchema}"`);
		await ensureMigrationsTable(db$1, options);
		const [migrations, runNames] = await Promise.all([loadMigrations(db$1, options, logger), getRunMigrations(db$1, options)]);
		if (options.checkOrder !== false) checkOrder(runNames, migrations);
		const toRun = getMigrationsToRun(options, runNames, migrations);
		if (toRun.length === 0) {
			logger.info("No migrations to run!");
			return [];
		}
		logger.info("> Migrating files:");
		for (const m of toRun) logger.info(`> - ${m.name}`);
		if (options.fake) await runMigrations(toRun, "markAsRun", options.direction);
		else if (options.singleTransaction) {
			await db$1.query("BEGIN");
			try {
				await runMigrations(toRun, "apply", options.direction);
				await db$1.query("COMMIT");
			} catch (error) {
				logger.warn("> Rolling back attempted migration ...");
				await db$1.query("ROLLBACK");
				throw error;
			}
		} else await runMigrations(toRun, "apply", options.direction);
		return toRun.map((m) => ({
			path: m.path,
			name: m.name,
			timestamp: m.timestamp
		}));
	} finally {
		if (db$1.connected()) {
			if (!options.noLock) await unlock(db$1, options.lockValue).catch((error) => {
				logger.warn(error.message);
			});
			await db$1.close();
		}
	}
}
//#endregion
export { PG_MIGRATE_LOCK_ID, loadMigrations, runner };
