// node-pg-migrate/dist/legacy/utils/comparators.js
import { getNumericPrefix } from "./fileNameUtils.js";
//#region src/utils/comparators.ts
function localeCompareStringsNumerically(a, b) {
	return a.localeCompare(b, void 0, {
		usage: "sort",
		numeric: true,
		sensitivity: "variant",
		ignorePunctuation: true
	});
}
/**
* Compares two file names by their numeric prefix.
* 17 digit numbers are interpreted as UTC date and converted to the number
* representation of that date. 1...4 digit numbers are interpreted as index
* based naming scheme.
*
* @param a the first file name to compare
* @param b the second file name to compare
* @param logger the logger to use for logging
* @returns a negative value if `a` is less than `b`, 0 if they are equal, and a positive value if `a` is greater than `b`
*/
function compareFileNamesByTimestamp(a, b, logger) {
	return getNumericPrefix(a, logger) - getNumericPrefix(b, logger);
}
/**
* Compares two file names by their numeric prefix and locale.
*
* @param a the first file name to compare
* @param b the second file name to compare
* @param logger the logger to use for logging
* @returns a negative value if `a` is less than `b`, 0 if they are equal, and a positive value if `a` is greater than `b`
*/
function compareMigrationFileNames(a, b, logger) {
	return compareFileNamesByTimestamp(a, b, logger) || localeCompareStringsNumerically(a, b);
}
//#endregion
export { compareFileNamesByTimestamp, compareMigrationFileNames, localeCompareStringsNumerically };
