// node-pg-migrate/dist/legacy/operations/indexes/shared.js
import { isPgLiteral } from "../../utils/PgLiteral.js";
import { isNameObject, isSchemaNameObject } from "../generalTypes.js";
import "../../utils/index.js";
//#region src/operations/indexes/shared.ts
function isIndexColumn(value) {
	return typeof value === "object" && value !== null && "name" in value && typeof value.name === "string";
}
function generateIndexName(table, columns, options, schemalize) {
	if (options.name) return isSchemaNameObject(table) ? {
		schema: table.schema,
		name: options.name
	} : options.name;
	const cols = columns.map((col, idx) => {
		if (isIndexColumn(col)) return schemalize(col.name);
		if (isPgLiteral(col)) {
			const literalValue = "value" in col ? col.value : String(col);
			throw new Error(`Index name must be provided when using PgLiteral columns (column #${idx + 1}: ${literalValue})`);
		}
		return schemalize(col);
	}).join("_");
	const uniq = "unique" in options && options.unique ? "_unique" : "";
	return isNameObject(table) ? {
		schema: table.schema,
		name: `${table.name}_${cols}${uniq}_index`
	} : `${table}_${cols}${uniq}_index`;
}
function generateColumnString(column, mOptions) {
	if (isPgLiteral(column)) return column.toString();
	const name = mOptions.schemalize(column);
	if (!/[^\w".]/.test(name)) return mOptions.literal(name);
	return /\)$/.test(name) ? name : `(${name})`;
}
function generateColumnsString(columns, mOptions) {
	return columns.map((column) => {
		if (typeof column === "string" || isPgLiteral(column)) return generateColumnString(column, mOptions);
		return [
			generateColumnString(column.name, mOptions),
			column.opclass ? mOptions.literal(column.opclass) : void 0,
			column.sort
		].filter((s) => typeof s === "string" && s !== "").join(" ");
	}).join(", ");
}
//#endregion
export { generateColumnString, generateColumnsString, generateIndexName };
